### PR TITLE
Initial work on a unit-testable version of the sync command.

### PR DESCRIPTION
**Pull request by bwbeach**
The new code is not used yet, but I think it's a good time to merge.

I'm working towards a sync implementation that can be multi-threaded, and that has clean layering so that it's testable.  The work so far includes folder classes (local and B2) that have a common interface, and a `make_sync_folder_sync_actions` function that figures out what needs to happen to get two folders in sync.

### CODE REVIEW

File: b2/b2.py [hunk 1]
```diff
@@ -1574,6 +1574,308 @@ def download_file_by_id_helper(
...
+    """
+    Holds information about one version of a file:
+
+       id - The B2 file id, or the local full path name
+       mod_time - modification time, in seconds
+       action - "hide" or "upload" (never "start")
+    """
+
+    def __init__(self, id, mod_time, action):
+        self.id = id
```
Comment (ppolewicz) [inline b2/b2.py]: `id` is a special word in Python. `id_` would be used by convention.

File: b2/b2.py [hunk 2]
```diff
@@ -1574,6 +1574,308 @@ def download_file_by_id_helper(
...
+            current_a = next_or_none(iter_a)
+        elif current_a.name < current_b.name:
+            yield (current_a, None)
+            current_a = next_or_none(iter_a)
+        elif current_b.name < current_a.name:
+            yield (None, current_b)
+            current_b = next_or_none(iter_b)
+        else:
+            assert current_a.name == current_b.name
+            yield (current_a, current_b)
```
Comment (ppolewicz) [inline b2/b2.py]: This statement can be moved to first line of the loop and other `yield` can be deleted

File: test/test_sync.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        shutil.rmtree(self.dirpath)
+        return None  # do not hide exception
+
+
+class TestLocalFolder(unittest.TestCase):
+    def test_slash_sorting(self):
+        # '/' should sort between '.' and '0'
+        with TempDir() as dir:
```
Comment (ppolewicz) [inline test/test_sync.py]: `dir` is a special word in Python

File: test/test_sync.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+            create_files(dir, ['hello.', 'hello/a', 'hello/b', 'hello0'])
+            folder = LocalFolder(dir)
+            files = list(folder.all_files())
+            names = list(map(lambda f: f.name, files))
```
Comment (ppolewicz) [inline test/test_sync.py]: `names = [f.name for f in fields]`

File: b2/b2.py (hunk 1 above)
Comment (bwbeach) [inline b2/b2.py]: right.

File: b2/b2.py (hunk 2 above)
Comment (bwbeach) [inline b2/b2.py]: I don't think so.  In the case where `current_a` has file `a.txt` and `current_b` has `b.txt`, it needs to yield `(current_a, None)`

File: test/test_sync.py (hunk 3 above)
Comment (bwbeach) [inline test/test_sync.py]: I've been coding in Java too much.  :-(

File: test/test_sync.py (hunk 4 above)
Comment (bwbeach) [inline test/test_sync.py]: yes.

**Comment by bwbeach** [comment]
Thanks for the careful review.

**Pull request merged**